### ISSUE TITLE
rsyslog with OpenSSL 1.1.1: SSL_ERROR_SYSCALL in 'osslRecordRecv'

### ISSUE THREAD

**Issue by Whissi** [rsyslog/rsyslog#3305]
I tested rsyslog against OpenSSL 1.1.1 and got two test failures:

```

------------------------------------------------------------
Test: ./imtcp-tls-ossl-basic.sh
------------------------------------------------------------
testbench: TZ env var not set, setting it to UTC
config rstb_456882_49aedce4_.conf is:
     1	module(load="../plugins/imdiag/.libs/imdiag")
     2	global(inputs.timeout.shutdown="60000")
     3	$IMDiagListenPortFileName rstb_456882_49aedce4.imdiag.port
     4	$IMDiagServerRun 0
     5	
     6	:syslogtag, contains, "rsyslogd"  ./rstb_456882_49aedce4.started
     7	###### end of testbench instrumentation part, test conf follows:
     8	
     9	global(	defaultNetstreamDriverCAFile="./tls-certs/ca.pem"
    10		defaultNetstreamDriverCertFile="./tls-certs/cert.pem"
    11		defaultNetstreamDriverKeyFile="./tls-certs/key.pem"
    12	#	debug.whitelist="on"
    13	#	debug.files=["nsd_ossl.c", "tcpsrv.c", "nsdsel_ossl.c", "nsdpoll_ptcp.c", "dnscache.c"]
    14	)
    15	
    16	module(	load="../plugins/imtcp/.libs/imtcp"
    17		StreamDriver.Name="ossl"
    18		StreamDriver.Mode="1"
    19		StreamDriver.AuthMode="anon" )
    20	input(	type="imtcp"
    21		port="54035" )
    22	
    23	template(name="outfmt" type="string" string="%msg:F,58:2%\n")
    24	:msg, contains, "msgnum:" action(	type="omfile" 
    25						template="outfmt"
    26						file=`echo $RSYSLOG_OUT_LOG`)
rsyslogd: NOTE: RSYSLOG_DEBUG_TIMEOUTS_TO_STDERR activated
rstb_456882_49aedce4:.pid1 found, pid 129208
rsyslogd startup msg seen, pid  129208
imdiag port: 36653
starting run 1
Sending 10000 messages.

00010000 messages sent
runtime: 0.004
closeTLSSess: closing SSL Session ...
End of tcpflood Run
Shutting down instance 1
rsyslogd: SSL_ERROR_SYSCALL in 'osslRecordRecv' [v8.39.0]
rsyslogd: netstream session 0x7fde7968d2a0 from 127.0.0.1 will be closed due to error [v8.39.0]
imdiag[36653]: mainqueue empty
rsyslogd debug: info: trying to cooperatively stop input ../plugins/imdiag/.libs/imdiag, timeout 60000 ms
rsyslogd debug: info: trying to cooperatively stop input imtcp, timeout 60000 ms
wait on shutdown of 129208
scanf error in index i=9830
sequence error detected in rstb_456882_49aedce4.out.log
number of lines in file: 9830 rstb_456882_49aedce4.out.log
sorted data has been placed in error.log, first 10 lines are:
     1	00000000
     2	00000001
     3	00000002
     4	00000003
     5	00000004
     6	00000005
     7	00000006
     8	00000007
     9	00000008
    10	00000009
---last 10 lines are:
  9821	00009820
  9822	00009821
  9823	00009822
  9824	00009823
  9825	00009824
  9826	00009825
  9827	00009826
  9828	00009827
  9829	00009828
  9830	00009829
UNSORTED data, first 10 lines are:
     1	00000000
     2	00000001
     3	00000002
     4	00000003
     5	00000004
     6	00000005
     7	00000006
     8	00000007
     9	00000008
    10	00000009
---last 10 lines are:
  9821	00009820
  9822	00009821
  9823	00009822
  9824	00009823
  9825	00009824
  9826	00009825
  9827	00009826
  9828	00009827
  9829	00009828
  9830	00009829
FAIL imtcp-tls-ossl-basic.sh (exit status: 1)

FAIL: imtcp-tls-ossl-x509name
=============================
```
and
```

------------------------------------------------------------
Test: ./imtcp-tls-ossl-x509name.sh
------------------------------------------------------------
testbench: TZ env var not set, setting it to UTC
config rstb_461625_4c95d5f9_.conf is:
     1	module(load="../plugins/imdiag/.libs/imdiag")
     2	global(inputs.timeout.shutdown="60000")
     3	$IMDiagListenPortFileName rstb_461625_4c95d5f9.imdiag.port
     4	$IMDiagServerRun 0
     5	
     6	:syslogtag, contains, "rsyslogd"  ./rstb_461625_4c95d5f9.started
     7	###### end of testbench instrumentation part, test conf follows:
     8	
     9	global(	defaultNetstreamDriverCAFile="./tls-certs/ca.pem"
    10		defaultNetstreamDriverCertFile="./tls-certs/cert.pem"
    11		defaultNetstreamDriverKeyFile="./tls-certs/key.pem"
    12	#	debug.whitelist="on"
    13	#	debug.files=["nsd_ossl.c", "tcpsrv.c", "nsdsel_ossl.c", "nsdpoll_ptcp.c", "dnscache.c"]
    14	)
    15	
    16	module(	load="../plugins/imtcp/.libs/imtcp"
    17		StreamDriver.Name="ossl"
    18		StreamDriver.Mode="1"
    19		StreamDriver.AuthMode="x509/name"
    20		PermittedPeer=["/CN=rsyslog-client/OU=Adiscon GmbH/O=Adiscon GmbH/L=Grossrinderfeld/ST=BW/C=DE/DC=rsyslog.com","rsyslog.com"]
    21		)
    22	input(	type="imtcp"
    23		port="42677" )
    24	
    25	template(name="outfmt" type="string" string="%msg:F,58:2%\n")
    26	:msg, contains, "msgnum:" action(	type="omfile" 
    27						template="outfmt"
    28						file=`echo $RSYSLOG_OUT_LOG`)
rsyslogd: NOTE: RSYSLOG_DEBUG_TIMEOUTS_TO_STDERR activated
rstb_461625_4c95d5f9:.pid1 found, pid 129706
rsyslogd startup msg seen, pid  129706
imdiag port: 34911
starting run 1
Sending 10000 messages.

00010000 messages sent
runtime: 0.003
closeTLSSess: closing SSL Session ...
End of tcpflood Run
Shutting down instance 1
rsyslogd: SSL_ERROR_SYSCALL in 'osslRecordRecv' [v8.39.0]
rsyslogd: netstream session 0x7ff24088d2a0 from 127.0.0.1 will be closed due to error [v8.39.0]
imdiag[34911]: mainqueue empty
wait on shutdown of 129706
rsyslogd debug: info: trying to cooperatively stop input ../plugins/imdiag/.libs/imdiag, timeout 60000 ms
rsyslogd debug: info: trying to cooperatively stop input imtcp, timeout 60000 ms
scanf error in index i=8638
sequence error detected in rstb_461625_4c95d5f9.out.log
number of lines in file: 8638 rstb_461625_4c95d5f9.out.log
sorted data has been placed in error.log, first 10 lines are:
     1	00000000
     2	00000001
     3	00000002
     4	00000003
     5	00000004
     6	00000005
     7	00000006
     8	00000007
     9	00000008
    10	00000009
---last 10 lines are:
  8629	00008628
  8630	00008629
  8631	00008630
  8632	00008631
  8633	00008632
  8634	00008633
  8635	00008634
  8636	00008635
  8637	00008636
  8638	00008637
UNSORTED data, first 10 lines are:
     1	00000000
     2	00000001
     3	00000002
     4	00000003
     5	00000004
     6	00000005
     7	00000006
     8	00000007
     9	00000008
    10	00000009
---last 10 lines are:
  8629	00008628
  8630	00008629
  8631	00008630
  8632	00008631
  8633	00008632
  8634	00008633
  8635	00008634
  8636	00008635
  8637	00008636
  8638	00008637
FAIL imtcp-tls-ossl-x509name.sh (exit status: 1)
```

**Comment by Whissi** [comment]
imtcp-tls-ossl-basic.debug.log: [imtcp-tls-ossl-basic.debug.log.gz](https://github.com/rsyslog/rsyslog/files/2592815/imtcp-tls-ossl-basic.debug.log.gz)

**Comment by rgerhards** [comment]
Thx! That's most probably an incarnation of missing sync - see PR. I guess there are quite some more of these issues.

**Comment by rgerhards** [comment]
@Whissi sorry - even though written pretty obvious, I overlooked the SYSCALL error. But it doesn't hurt to have the test updated ;-)

@alorbach could you have a look?

**Comment by alorbach** [comment]
I did a fast research on that error. 
All I found is, that it is properly related to local network connection problems. Caused by firewall settings for example.

Can you make sure the problem is not caused due network problems locally?

**Comment by Whissi** [comment]
Network is running in a sandbox but 'localhost' network is unrestricted otherwise. So yes, I am pretty sure this is not caused by a firewall.

PS: Could you please re-open this issue? I think it got only closed because the PR mentioned above accidentally linked this issue but like @rgerhards noticed a few seconds after he created that link, the solved from problem the PR has nothing to do with this bug.

**Comment by rgerhards** [comment]
@Whissi sry for not re-opening

@alorbach is the error message the best we can get? e.g. is it the full openssl error stack?

**Comment by alorbach** [comment]
If there was an openssl error stack, it would be in the debug log. 
The code at this position will output this automatically.

**Comment by rgerhards** [comment]
Apologies for the very late follow-up. This may no longer help the original reporter, but I am recording the disposition for people finding this through search while working through the backlog. We do not close reports merely because they are old.

This concerned the newly released OpenSSL 1.1.1 path and rsyslog 8.39. Current rsyslog is routinely built and tested with modern OpenSSL, including dedicated OpenSSL send/receive, anonymous/certificate, SNI, chain, and error-path suites. The retained report has no current reproduction or OpenSSL error stack to identify a still-existing defect.

Closing as obsolete/not currently reproducible. If SSL_ERROR_SYSCALL occurs on a current release, please open a new issue with rsyslog/OpenSSL versions, both endpoint configs, the preceding OpenSSL error stack, and a packet capture.
